File: AspNetMvcUnleashed/0672329980_source/Chapter08/VB/MvcApplication1/MvcApplication1.Tests/Controllers/Movie2ControllerTests.vb
Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports System.Web.Mvc

<TestClass()> _
    Public Class Movie2ControllerTests

    <TestMethod()> _
    Public Sub DirectorRequired()
        ' Arrange
        Dim modelState = New ModelStateDictionary()
        Dim service = New MovieService(modelState, New FakeMovieRepository())
        Dim controller = New Movie2Controller(service)
        Dim movieToCreate = Movie.CreateMovie(0, "Star Wars", String.Empty, DateTime.Parse("1/1/1977"))

        ' Act
        controller.Create(movieToCreate)

        ' Assert
        Assert.IsTrue(HasErrorMessage(modelState("Director"), "Director is required."))
    End Sub


    Private Function HasErrorMessage(ByVal modelState As ModelState, ByVal errorMessage As String) As Boolean
        For Each modelError In modelState.Errors
            If modelError.ErrorMessage = errorMessage Then
                Return True
            End If
        Next
        Return False
    End Function


End Class

File: AspNetMvcUnleashed/0672329980_source/Chapter08/CS/MvcApplication1/MvcApplication1.Tests/Controllers/Movie2ControllerTests.cs
using System;
using System.Web.Mvc;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using MvcApplication1.Controllers;
using MvcApplication1.Models;
using MvcApplication1.Tests.Models;

namespace MvcApplication1.Tests.Controllers
{
    [TestClass]
    public class Movie2ControllerTests
    {

        [TestMethod]
        public void DirectorRequired()
        {
            // Arrange
            var modelState = new ModelStateDictionary();
            var service = new MovieService(modelState, new FakeMovieRepository());
            var controller = new Movie2Controller(service);
            var movieToCreate = Movie.CreateMovie(0, "Star Wars", String.Empty, DateTime.Parse("1/1/1977"));

            // Act
            controller.Create(movieToCreate);

            // Assert
            Assert.IsTrue(HasErrorMessage(modelState["Director"], "Director is required."));        
        }


        private bool HasErrorMessage(ModelState modelState, string errorMessage)
        {
            foreach (var error in modelState.Errors)
            {
                if (error.ErrorMessage == errorMessage)
                    return true;
            }
            return false;
        }


    }
}
